File: backend/utils/bigram.py
def _bigrams(text: str):
    chars = list(text.lower().replace(" ", "").replace("\t", "").replace("\n", ""))
    return set(chars[i] + chars[i + 1] for i in range(len(chars) - 1))


def bigram_score(query: str, text: str) -> float:
    if not query or not text:
        return 0.0
    q_grams = _bigrams(query)
    if not q_grams:
        return 0.0
    t_grams = _bigrams(text)
    hits = sum(1 for g in q_grams if g in t_grams)
    return hits / len(q_grams)

File: frontend/src/utils/bigram.js
export function bigrams(text) {
  const chars = [...text.toLowerCase().replace(/\s+/g, '')]
  const set = new Set()
  for (let i = 0; i < chars.length - 1; i++) set.add(chars[i] + chars[i + 1])
  return set
}

export function bigramScore(query, text) {
  if (!query || !text) return 0
  const qGrams = bigrams(query)
  if (!qGrams.size) return 0
  const tGrams = bigrams(text)
  let hits = 0
  for (const g of qGrams) if (tGrams.has(g)) hits++
  return hits / qGrams.size
}
